[RFC] Add documentation for :terminal and Terminal-mode

nelstrom opened the pull request:
Fixes #6756.

nelstrom committed:
Add documentation for `:terminal` with no arguments

nelstrom commented:
While we're at it, it might be worthwhile documenting what the bang in `:terminal!` does.

nelstrom committed:
Update `:h :te` docs to say that new term takes over current split

mhinz commented on runtime/doc/various.txt:
```diff
@@ -211,6 +211,11 @@ g8			Print the hex values of the bytes used in the
 :te[rminal][!]		Open an interactive shell in a |terminal-emulator|.
 			Use the shell specified in 'shell'.
 
+			The new terminal buffer takes over the current split
```
No such thing as a "split window". Just "window". :] And maybe the whole sentence can be left away:

> Creating the terminal buffer fails when changes have been made to the current buffer, unless 'hidden' is set.

nelstrom replied:
Ok, I'll `s/split window/window`

Happy to remove the stuff about failing if the current buffer has changes. I thought it was worthwhile at least having a go at documenting the meaning of `[!]` in `:terminal!`, but there's a lot of complexity there and to document it thoroughly would mean adding lots and lots of words.

nelstrom committed:
s/split window/window

mhinz commented on runtime/doc/various.txt:
```diff
@@ -211,10 +211,8 @@ g8			Print the hex values of the bytes used in the
-			window (like |:edit|). This fails when changes have
-			been made to the current buffer, unless 'hidden' is
-			set.
+			The new terminal buffer takes over the current window
```
Hmm, now you did the opposite of what I had in mind. :-P

I mean, a new buffer acts in the current window by default, so I thought it shouldn't be mentioned. But the remark about unsaved changes is quite important.

Moreover, as far as I recall, "taking over a window" is a wording that isn't used anywhere else in the docs, that's why I meant it should be left away or be merged with the following sentence:

Something like this:

```
Creating the terminal buffer fails when changes have been
made to the current buffer, unless 'hidden' is set.
```

FWIW, usually Justin chimes in for the final wording, so maybe wait a moment before you change it yet again. :-)

justinmk commented on runtime/doc/various.txt:
```diff
@@ -208,13 +208,20 @@ g8			Print the hex values of the bytes used in the
+
+						  *:terminal_c*
 :te[rminal][!] {cmd}	Execute {cmd} with 'shell' in a |terminal-emulator|
-                        buffer.  Equivalent to: >
```
Can't we just add brackets around `{cmd}`? 

    :te[rminal][!] [{cmd}]	Execute {cmd} with 'shell' in a |terminal-emulator|

Also like @mhinz implied, the "take over" is already implied by the words "in a ~~new~~ terminal buffer".  (We should add "new" to that sentence.) 

Users presumably are not surprised when `:edit` "takes over" a window so they should not be surprised by `:terminal`. The reason people _are_ surprised by `:terminal` is because they expect it to behave like `:!`. And IMO `:terminal` _should_ split the window, but it might be too late to change that behavior now....

nelstrom commented:
I'll have another go at rewriting this after sleeping on it.

nelstrom committed:
Combine :terminal_c and :terminal help entries back together again

Use [{cmd}] to indicate that the {cmd} is optional

nelstrom committed:
Mention that :terminal will fail if current buffer has been modified

...unless 'hidden' is set.

nelstrom committed:
Add helptag for *terminal-buffer*

It feels clunky to say "a new |terminal-emulator| buffer"

nelstrom committed:
Add an entry about |Terminal-mode| under "Modes, introduction"

nelstrom committed:
Add a line about `:terminal` without {cmd}

nelstrom committed:
s/See jobstart()/See termopen()

If we read the entry about `termopen()`, it gives us some context and
then refers us on to the entry about `jobstart()` if we still want to
know more.

justinmk commented on runtime/doc/nvim_terminal_emulator.txt:
```diff
@@ -5,6 +5,7 @@
 
 
 Terminal emulator					   *terminal-emulator*
+							     *terminal-buffer*
```
I changed this to just `*terminal*`. In a pending PR I cleaned up all user docs so that the word "terminal" is always understand to mean the "built-in embedded terminal", and "TUI" is used in the few cases where we need to discuss external terminals.

nelstrom replied:
That's a nice solution. I'm not sure that I would automatically understand TUI to mean that, but if you specify that that's the way it's intended it should work.

nelstrom replied:
That's a nice solution. I'm not sure that I would automatically understand TUI to mean that, but if you specify that that's the way it's intended it should work.

justinmk commented:
Merged, thanks!
